group-coordinator/src/test/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroupTest.java

```diff
@@ -0,0 +1,500 @@
+            .build();
+        consumerGroup.updateMember(member2);
+
+        assertEquals(ConsumerGroup.ConsumerGroupState.STABLE, consumerGroup.state());
+
+        consumerGroup.removeMember("member1");
+        consumerGroup.removeMember("member2");
+
+        assertEquals(ConsumerGroup.ConsumerGroupState.EMPTY, consumerGroup.state());
+    }
+
+    @Test
+    public void testPreferredServerAssignor() {
+        ConsumerGroup consumerGroup = createConsumerGroup("foo");
+
+        ConsumerGroupMember member1 = new ConsumerGroupMember.Builder("member1")
+            .setServerAssignorName("range")
+            .build();
+        ConsumerGroupMember member2 = new ConsumerGroupMember.Builder("member2")
+            .setServerAssignorName("range")
+            .build();
+        ConsumerGroupMember member3 = new ConsumerGroupMember.Builder("member3")
+            .setServerAssignorName("uniform")
+            .build();
+
+        assertEquals(
+            Optional.empty(),
+            consumerGroup.preferredServerAssignor(null, null)
+        );
+
+        assertEquals(
+            Optional.of("range"),
+            consumerGroup.preferredServerAssignor(null, member1)
+        );
+
+        consumerGroup.updateMember(member1);
+
+        assertEquals(
+            Optional.of("range"),
+            consumerGroup.preferredServerAssignor(null, null)
+        );
+
+        assertEquals(
+            Optional.empty(),
+            consumerGroup.preferredServerAssignor(member1, null)
+        );
+
+        assertEquals(
+            Optional.of("range"),
+            consumerGroup.preferredServerAssignor(null, member2)
+        );
+
+        consumerGroup.updateMember(member2);
+
+        assertEquals(
+            Optional.of("range"),
+            consumerGroup.preferredServerAssignor(null, null)
+        );
+
+        consumerGroup.updateMember(member3);
+
+        assertEquals(
+            Optional.of("range"),
+            consumerGroup.preferredServerAssignor(null, null)
+        );
+
+        // Members without assignors
+        ConsumerGroupMember updatedMember1 = new ConsumerGroupMember.Builder("member1")
+            .setServerAssignorName(null)
+            .build();
+        ConsumerGroupMember updatedMember2 = new ConsumerGroupMember.Builder("member2")
+            .setServerAssignorName(null)
+            .build();
+        ConsumerGroupMember updatedMember3 = new ConsumerGroupMember.Builder("member3")
+            .setServerAssignorName(null)
+            .build();
+
+
+        Optional<String> assignor = consumerGroup.preferredServerAssignor(member1, updatedMember1);
+        assertTrue(assignor.equals(Optional.of("range")) || assignor.equals(Optional.of("uniform")));
+
+        consumerGroup.updateMember(updatedMember1);
+
+        assignor = consumerGroup.preferredServerAssignor(member1, updatedMember1);
+        assertTrue(assignor.equals(Optional.of("range")) || assignor.equals(Optional.of("uniform")));
```

Comment:
We have member 2 as range and member 3 as uniform so that's why it is 50/50

Reply:
i'm confused because i was referring to member 2 here:
range: 1 (member 2)
uniform: 1 (member 1)

after `consumerGroup.updateMember(updatedMember1);` in L315.

in L317, we call `preferredServerAssignor()` which gets a copy, then will decrement the count of "range" since member1 (range) is passed in as the oldMember argument. So shouldn't the count for "range" be 0?

i think i'm missing something

Reply:
Updated the test. I had a mistake in these.